### PR TITLE
Additional config options for hot reloader web socket connection

### PR DESCRIPTION
**Pull request by jaslakson**
We run Next on a sub-set of urls and the root url is not included in this group. Additionally, we run `nginx` over https as a proxy server for all urls, in all environments (local, staging, production, etc). The existing WebSocket configuration options are not sufficient for our needs.

Proposed changes:

* conditionally set WebSocket protocol based on location.protocol (`wss` or `ws`)
* add the following configuration options for `onDemandEntries`
  * `websocketProxyPort`: allows us to configure `nginx` to listen on one port and proxy pass to `websocketProxyPort`
  * `websocketProxyPath`: since the root of our site is not served by Next, we need a path to use in order to establish the socket connection

Tests:

I might need a little help here - looking at the existing `ondemand` tests, I wasn't sure where or how to test these options or if testing these options is completely necessary.

Thanks for all of your hard work on Next.js. My WebSocket experience is limited so I am open to feedback or pointers on how to do this differently. 

I tried the steps outlined in your contributing guide but ran into errors when trying to test this in my local repository.

### CODE REVIEW

File: packages/next-server/server/config.js [hunk 1]
```diff
@@ -18,7 +18,9 @@ const defaultConfig = {
   onDemandEntries: {
     maxInactiveAge: 60 * 1000,
     pagesBufferLength: 2,
-    websocketPort: 0
+    websocketPort: 0,
+    websocketProxyPath: null,
```
Comment (timneutkens) [inline packages/next-server/server/config.js]: Should this instead just be taking `assetPrefix` into account instead?

**Comment by timneutkens** [comment]
>   /home/circleci/repo/packages/next/server/hot-reloader.js:170:6: Extra semicolon.

Looks like linting didn't run.

File: packages/next-server/server/config.js (hunk 1 above)
Comment (jaslakson) [reply packages/next-server/server/config.js]: I don't think so - at least in my case what I'm trying to accomplish here is set a unique path which corresponds to a location block in my nginx config. This allows the `sw` request to successfully establish itself. 

Here's a snippet from my conf files so you can see what I am working with:

I am listening on `7001` and forwarding to `7002`, which I have set as my `websocketPort` locally.

```
location /hmr {
  proxy_http_version 1.1;
  proxy_set_header Upgrade $http_upgrade;
  proxy_set_header Connection "upgrade";
  proxy_pass http://espresso_hot_server;
}
...elsewhere
upstream espresso_hot_server {
  server 127.0.0.1:7002;
}
```

File: packages/next/client/on-demand-entries-client.js [hunk 2]
```diff
@@ -19,7 +22,7 @@ export default async ({ assetPrefix }) => {
     }
 
     return new Promise(resolve => {
-      ws = new WebSocket(`ws://${hostname}:${process.env.NEXT_WS_PORT}`)
+      ws = new WebSocket(`${wsProtocol}://${hostname}:${wsPort}${wsPath ? `/${wsPath}` : ''}`)
```
Comment (timneutkens) [inline packages/next/client/on-demand-entries-client.js]:
```suggestion
      ws = new WebSocket(`${wsProtocol}://${hostname}:${wsPort}${wsPath}`)
```

File: packages/next/client/on-demand-entries-client.js [hunk 3]
```diff
@@ -3,7 +3,10 @@
 import Router from 'next/router'
 import fetch from 'unfetch'
 
-const { hostname } = location
+const { hostname, protocol } = location
+const wsProtocol = protocol.includes('https') ? 'wss' : 'ws'
+const wsPort = process.env.NEXT_WS_PROXY_PORT || process.env.NEXT_WS_PORT
+const wsPath = process.env.NEXT_WS_PROXY_PATH || null
```
Comment (timneutkens) [inline packages/next/client/on-demand-entries-client.js]:
```suggestion
const wsPath = process.env.NEXT_WS_PROXY_PATH || '/'
```

File: packages/next-server/server/config.js (hunk 1 above)
Comment (timneutkens) [inline packages/next-server/server/config.js]:
```suggestion
    websocketProxyPath: '/',
```

File: packages/next-server/server/config.js [hunk 4]
```diff
@@ -18,7 +18,9 @@ const defaultConfig = {
...
+    websocketProxyPort: null
```
Comment (timneutkens) [inline packages/next-server/server/config.js]: Why have another port next to this one? 🤔

Comment (jaslakson) [reply packages/next-server/server/config.js]: I reserve `7001` on `nginx` and `7002` in Next. The two can't be the same port.

Comment (timneutkens) [reply packages/next-server/server/config.js]: Ah now I get it, sounds good to me 💯

**Comment by timneutkens** [comment]
Could you add these new options to the readme too? After that + the changes I proposed I can merge the PR 👍

File: packages/next/server/hot-reloader.js [hunk 5]
```diff
@@ -164,10 +164,18 @@ export default class HotReloader {
...
-  addWsPort (configs) {
-    configs[0].plugins.push(new webpack.DefinePlugin({
+  addWsConfig (configs) {
+    const opts = {
       'process.env.NEXT_WS_PORT': this.wsPort
-    }))
+    }
+    const { websocketProxyPath, websocketProxyPort } = this.config.onDemandEntries
+    if (websocketProxyPath) {
+      opts['process.env.NEXT_WS_PROXY_PATH'] = websocketProxyPath
```
Comment (jaslakson) [inline packages/next/server/hot-reloader.js]: @timneutkens Do you think I should change this line too now that we've updated `on-demand-entries-client` - i.e. would it be better to always define `process.env.NEXT_WS_PROXY_PATH` and `process.env.NEXT_WS_PROXY_PORT` based on their default or user defined values? As it stands right now, I think `wsPath` would add `undefined` to the end of the socket path.

Comment (timneutkens) [reply packages/next/server/hot-reloader.js]: I was thinking the same thing, I just pushed up a commit to both fix the setting / make test pass + change this!

**Pull request merged**